### PR TITLE
feat(relay): add Relay support

### PR DESCRIPTION
**Pull request by damassi**
This PR adds Relay support and adds an example app under `/pages/artist/[artistSlug]`: 
```
http://localhost:3000/artist/andy-warhol
```

The new dev flow: `yarn dev` will run `yarn relay` and `next dev` concurrently, as we do in force.  

> ⚠️ Once this is merged, be sure to run `yarn clean` if you've booted the app before locally. 

### Creating New Pages

What we're now doing at the framework level is making a request out to MP during the server pass and then rehydrating the relay store, which is then passed over to the client. As such we have a new helper function to make that request in `next`s `getServerSideProps` function:

```tsx
// pages/artist/[artistSlug].tsx

export default function Artist({ artist }: ArtistProps) {
 return <ArtistName artist={artist} /> 
}

export const getServerSideProps: GetServerSideProps = async (context) => {
  const props = await fetchRelayData({
    query: graphql`
      query ArtistQuery($artistSlug: String!) {
        artist(id: $artistSlug) {
          ...ArtistName_artist
          bio
        }
      }
    `,
    variables: {
      artistSlug: context.query.artistSlug,
    },
    cache: true,
  })

  return {
    props,
  }
}
```

Once one has set up their root next.js page query like the above, all nested sub-components can take advantage of relay's new hooks functionality:

```tsx
interface ArtistNameProps {
  artist: ArtistName_artist$key
}

const ArtistName: React.FC<ArtistNameProps> = ({ artist }) => {
  const data = useFragment(
    graphql`
      fragment ArtistName_artist on Artist {
        name
      }
    `,
    artist
  )

  return (
    <Box>
      <Text variant="xl">{data?.name}</Text>
    </Box>
  )
}
```

### Generating Relay Queries 

Relay queries **must** be unique, which in turn means that all relay-emitted types must be unique. This impacts some of the conventionally minimalist idioms around dynamic path file-naming in next ("[slug].tsx"). And since Relay looks at the file name to generate the type, by extension these file names must also be unique. 

So given the below file structure:

```
pages/
  artists/
     dedupe/
```

what I would suggest is using each folder name as a prefix for the dynamic path at the leaf:

```
pages/
  artists/
     dedupe/
       [artistsDedupeSlug].tsx
```

### Additional changes include

- type-checking on pre-push
- stubs out files for future relay testing support 
- renames `lib` -> `system` to better describe the purpose ("lib" is so vague) and to match force. The `system` folder now contains `auth` and `relay`
- stubs out a react error boundary 

Screen Shot 2022-03-25 at 3 55 58 PM

### CODE REVIEW

File: src/pages/artist/ArtistName.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { Text, Box } from "@artsy/palette"
+import { graphql, useFragment } from "react-relay"
+import { ArtistName_artist$key } from "__generated__/ArtistName_artist.graphql"
```
Comment (anandaroop) [inline src/pages/artist/ArtistName.tsx:3]: **question:** Why are we importing this `$key` here? In Force I believe it's customary to just import an ordinary looking fragment name, like `ArtistName_artist`, no?

File: src/pages/artist/ArtistName.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+const ArtistName: React.FC<ArtistNameProps> = ({ artist }) => {
+  const data = useFragment(
+    graphql`
+      fragment ArtistName_artist on Artist {
+        name
+      }
+    `,
+    artist
+  )
```
Comment (anandaroop) [inline src/pages/artist/ArtistName.tsx:17]: This is certainly an improvement

**Review by anandaroop — COMMENTED** [review]
Thanks so much for adding this 🙏🏽, this will be a huge unblock.

I regret not asking for more granular commits — I think it would be _very_ instructive to see the step-by-step, since this has proved to be anything but "zero-config" 😅

The description and example app are very helpful, and gtk about the added caveats re slugs etc.

I only have one suggested change, and then a few questions for my own understanding...

File: package.json [hunk 3]
```diff
@@ -5,29 +5,28 @@
...
     "clean": "rm -rf .next && yarn dev",
-    "dev": "next dev",
+    "dev": "concurrently 'yarn relay --watch' 'next dev'",
     "format": "prettier --write \"{pages,components,lib}/**/*.ts{,x}\"",
     "lint": "next lint --fix",
     "prepare": "husky install",
+    "relay": "relay-compiler",
     "start": "next start",
     "storybook": "start-storybook -p 6006",
+    "sync-schema": "scripts/sync-schema.sh",
```
[the remark this answers is not in the document]
Comment (damassi) [reply package.json:15]: We'll want to add `admin-next` here: https://github.com/artsy/metaphysics/blob/main/scripts/push-schema-changes.js#L64-L77

File: src/system/relay/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  urlMiddleware,
+} from "react-relay-network-modern"
+
+const { publicRuntimeConfig } = getConfig()
+const isClient = typeof window !== "undefined"
+
+export const middleware: Array<any> = [
+  urlMiddleware({
+    url: publicRuntimeConfig.NEXT_PUBLIC_METAPHYSICS_URL! + "/v2",
+  }),
```
[the remark this answers is not in the document]
Comment (damassi) [reply src/system/relay/middleware.ts:14]: Yup, exactly. Disabling JS to check out SSR (during localhost dev), it works well.

File: src/system/relay/setupTestWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  typeof import("@testing-library/dom/types/queries"),
+  HTMLElement
+>
+
+type RenderWithRelay = RTLRenderResult & { env: RelayMockEnvironment }
+
+export const setupTestWrapper = <T extends OperationType>({
+  Component,
+  query,
+  variables = {},
```
[the remark this answers is not in the document]
Comment (damassi) [reply src/system/relay/setupTestWrapper.tsx:36]: This code _does_ work, but not quite yet in the app. I still need to push forward a bit on compiling `graphql` tags via `next/babel` (from within Jest).

**Comment by damassi** [comment]
@anandaroop 

> since this has proved to be anything but "zero-config"

I was mistaken about that in the sense that the old `babel-plugin-relay` is now a part of Relay's core, but there's still setup involved with the GraphQL library itself. Same thing with styled-components. This is part of their babel -> SWC push.

**Pull request merged**